File: backup.py
#!/usr/bin/env python3
"""
Move any files in Templates_docx, Templates_markdown, or Templates_txt that are
not listed in reports_index.json into backup/, preserving folder structure.
"""

import json
import shutil
from pathlib import Path
from typing import Dict, Iterable, Set

BASE_DIR = Path(__file__).parent
TARGETS: Dict[str, str] = {
    "Templates_docx": "*.docx",
    "Templates_markdown": "*.md",
    "Templates_txt": "*.txt",
}
INDEX_PATH = BASE_DIR / "reports_index.json"
BACKUP_DIR = BASE_DIR / "backup"


def load_index() -> Dict[str, Set[str]]:
    if not INDEX_PATH.exists():
        raise SystemExit("Index file not found. Run generate_index.py first.")

    data = json.loads(INDEX_PATH.read_text(encoding="utf-8"))
    indexed: Dict[str, Set[str]] = {}
    for folder_name in TARGETS.keys():
        indexed[folder_name] = set(data.get(folder_name, []))
    return indexed


def iter_files(folder: Path, pattern: str) -> Iterable[Path]:
    return (p for p in folder.glob(pattern) if p.is_file())


def move_unindexed(indexed: Dict[str, Set[str]]) -> None:
    BACKUP_DIR.mkdir(exist_ok=True)
    moved = 0
    for folder_name, pattern in TARGETS.items():
        folder = BASE_DIR / folder_name
        if not folder.exists():
            print(f"Skipping missing folder: {folder}")
            continue

        expected = indexed.get(folder_name, set())
        for path in iter_files(folder, pattern):
            rel_path = str(path.relative_to(BASE_DIR))
            if rel_path in expected:
                continue

            dest = BACKUP_DIR / rel_path
            dest.parent.mkdir(parents=True, exist_ok=True)
            if dest.exists():
                print(f"Skip {rel_path}: destination already exists in backup.")
                continue

            shutil.move(str(path), str(dest))
            moved += 1
            print(f"Moved {rel_path} -> {dest.relative_to(BASE_DIR)}")

    print(f"\nDone. Files moved: {moved}")


def main() -> None:
    indexed = load_index()
    move_unindexed(indexed)


if __name__ == "__main__":
    main()

File: rust_converters/src/bin/backup.rs
use anyhow::{anyhow, Result};
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::{Path, PathBuf};

const TARGETS: [(&str, &str); 3] = [
    ("Templates_docx", "docx"),
    ("Templates_markdown", "md"),
    ("Templates_txt", "txt"),
];

type IndexMap = HashMap<String, Vec<String>>;

fn load_index(root: &Path) -> Result<IndexMap> {
    let path = root.join("reports_index.json");
    if !path.exists() {
        return Err(anyhow!(
            "Index file not found at {}. Run generate_index first.",
            path.display()
        ));
    }
    let contents = fs::read_to_string(&path)?;
    let parsed: IndexMap = serde_json::from_str(&contents)?;
    Ok(parsed)
}

fn should_keep(path: &Path, expected: &HashSet<String>, root: &Path) -> bool {
    let rel = path
        .strip_prefix(root)
        .unwrap_or(path)
        .to_string_lossy()
        .to_string();
    expected.contains(&rel)
}

fn move_unindexed(root: &Path, index: &IndexMap) -> Result<usize> {
    let backup_dir = root.join("backup");
    fs::create_dir_all(&backup_dir)?;
    let mut moved = 0usize;

    for (folder, ext) in TARGETS {
        let dir = root.join(folder);
        if !dir.exists() {
            eprintln!("Skipping missing folder: {}", dir.display());
            continue;
        }

        let expected: HashSet<String> = index
            .get(folder)
            .cloned()
            .unwrap_or_default()
            .into_iter()
            .collect();

        for entry in fs::read_dir(&dir)? {
            let path = entry?.path();
            if !path.is_file() {
                continue;
            }
            let has_ext = path
                .extension()
                .and_then(|s| s.to_str())
                .map(|s| s.eq_ignore_ascii_case(ext))
                .unwrap_or(false);
            if !has_ext {
                continue;
            }
            if should_keep(&path, &expected, root) {
                continue;
            }

            let rel = path.strip_prefix(root).unwrap_or(&path);
            let dest = backup_dir.join(rel);
            if dest.exists() {
                eprintln!(
                    "Skip {}: destination already exists",
                    rel.to_string_lossy()
                );
                continue;
            }
            if let Some(parent) = dest.parent() {
                fs::create_dir_all(parent)?;
            }

            fs::rename(&path, &dest)?;
            moved += 1;
            println!(
                "Moved {} -> {}",
                rel.to_string_lossy(),
                dest.strip_prefix(root)
                    .unwrap_or(&dest)
                    .to_string_lossy()
            );
        }
    }

    Ok(moved)
}

fn main() -> Result<()> {
    let root = PathBuf::from(".");
    let index = load_index(&root)?;
    let moved = move_unindexed(&root, &index)?;
    println!("\nDone. Files moved: {}", moved);
    Ok(())
}
